internal/querycoordv2/meta/target_manager.go

```diff
@@ -3,184 +3,278 @@ package meta
+	// all remove segment/channel operation happens on Both current and next -> delete status should be consistent
+	current *target
+	next    *target
 }
 
 func NewTargetManager() *TargetManager {
 	return &TargetManager{
-		segments:   make(map[int64]*datapb.SegmentInfo),
-		dmChannels: make(map[string]*DmChannel),
+		current: newTarget(),
+		next:    newTarget(),
 	}
 }
 
+func (mgr *TargetManager) UpdateCollectionCurrentTarget(collectionID int64, partitionIDs ...int64) {
+	mgr.rwMutex.Lock()
+	defer mgr.rwMutex.Unlock()
+	log.Info("start to update current target for collection",
+		zap.Int64("collectionID", collectionID),
+		zap.Int64s("PartitionIDs", partitionIDs))
+
+	newTarget := mgr.next.getCollectionTarget(collectionID)
+	if newTarget == nil || newTarget.isEmpty() {
+		log.Info("next target does not exist, skip it",
+			zap.Int64("collectionID", collectionID),
+			zap.Int64s("PartitionIDs", partitionIDs))
+		return
+	}
+	mgr.current.updateCollectionTarget(collectionID, newTarget)
+	mgr.next.removeCollectionTarget(collectionID)
+
+	log.Info("finish to update current target for collection",
+		zap.Int64("collectionID", collectionID),
+		zap.Int64s("PartitionIDs", partitionIDs),
+		zap.Int64s("segments", newTarget.GetAllSegmentIDs()),
+		zap.Strings("channels", newTarget.GetAllDmChannelNames()))
+}
+
+func (mgr *TargetManager) UpdateCollectionNextTarget(collectionID int64, newTarget *CollectionTarget) {
+	mgr.rwMutex.Lock()
+	defer mgr.rwMutex.Unlock()
+	log.Info("start to update next targets for collection",
+		zap.Int64("collectionID", collectionID))
+
+	mgr.next.updateCollectionTarget(collectionID, newTarget)
+
+	log.Info("finish to update next targets for collection",
+		zap.Int64("collectionID", collectionID),
+		zap.Int64s("segments", newTarget.GetAllSegmentIDs()),
+		zap.Strings("channels", newTarget.GetAllDmChannelNames()))
+}
+
 // RemoveCollection removes all channels and segments in the given collection
 func (mgr *TargetManager) RemoveCollection(collectionID int64) {
-	mgr.rwmutex.Lock()
-	defer mgr.rwmutex.Unlock()
+	mgr.rwMutex.Lock()
+	defer mgr.rwMutex.Unlock()
+	log.Info("remove collection from targets",
+		zap.Int64("collectionID", collectionID))
 
-	log.Info("remove collection from targets")
-	for _, segment := range mgr.segments {
-		if segment.CollectionID == collectionID {
-			mgr.removeSegment(segment.GetID())
-		}
-	}
-	for _, dmChannel := range mgr.dmChannels {
-		if dmChannel.CollectionID == collectionID {
-			mgr.removeDmChannel(dmChannel.GetChannelName())
-		}
-	}
+	mgr.current.removeCollectionTarget(collectionID)
+	mgr.next.removeCollectionTarget(collectionID)
 }
 
 // RemovePartition removes all segment in the given partition,
 // NOTE: this doesn't remove any channel even the given one is the only partition
-func (mgr *TargetManager) RemovePartition(partitionID int64) {
-	mgr.rwmutex.Lock()
-	defer mgr.rwmutex.Unlock()
-
+func (mgr *TargetManager) RemovePartition(collectionID int64, partitionIDs ...int64) {
```

Comment:
It seems that this method can't guarantee the next of next target won't contain the removed partitions, if the target observer is running at line 111, and this method is running just to the last line, the new next target will contain the removed partitions